File: Boj-Kotlin&Java/백준10942번팰린드롬.java
package 백준;

import java.io.*;
import java.util.StringTokenizer;

public class 백준10942번팰린드롬 {
    static int n;
    static int[] arr;
    static StringTokenizer st;
    static boolean[][] d;
    public static void main(String[] args) throws IOException {
        BufferedReader br=new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw=new BufferedWriter(new OutputStreamWriter(System.out));
        n=Integer.parseInt(br.readLine());
        st=new StringTokenizer(br.readLine());
        arr=new int[n];
        d=new boolean[n][n];
        for(int i=0; i<n; i++)
            arr[i]=Integer.parseInt(st.nextToken());
        for(int i=0; i<n; i++)
            d[i][i]=true;
        for(int i=0; i<n-1; i++){
            if(arr[i]==arr[i+1])
                d[i][i+1]=true;
        }
        for (int k=2; k<n; k++) {
            for (int i=0; i<n-k; i++) {
                int j = i+k;
                if (arr[i] == arr[j] && d[i+1][j-1]) {
                    d[i][j] = true;
                }
            }
        }
        int m=Integer.parseInt(br.readLine());
        for(int i=0; i<m; i++)
        {
            st=new StringTokenizer(br.readLine());
            int a=Integer.parseInt(st.nextToken());
            int b=Integer.parseInt(st.nextToken());
            if(d[a-1][b-1])
                bw.write("1\n");
            else
                bw.write("0\n");
        }
        bw.flush();
    }
}

File: Boj-Kotlin&Java/백준 10942번 팰린드롬.kt
fun main(args: Array<String>) {
    val br=System.`in`.bufferedReader()
    val bw=System.out.bufferedWriter()
    val n=br.readLine().toInt()
    val arr=br.readLine().split(" ").map { it.toInt() }
    val dp=Array(n){Array(n){false}}
    for(i in 0 until n)
        dp[i][i]=true
    for(i in 0 until n-1)
    {
        if(arr[i]==arr[i+1])
            dp[i][i+1]=true
    }
    for(i in 2 until n)
    {
        for(j in 0 until n-i)
        {
            val k=i+j
            if(arr[j]==arr[k] && dp[j+1][k-1])
                dp[j][k]=true
        }
    }
    val m=br.readLine().toInt()

    for(i in 0 until m)
    {
        val (a, b)=br.readLine().split(" ").map { it.toInt() }
        if(dp[a-1][b-1])
            bw.write("1\n")
        else
            bw.write("0\n")
    }
    bw.flush()
}